File: drivers/zig/src/redwire/scram.zig
// SCRAM-SHA-256 client primitives. Mirrors `src/auth/scram.rs` and
// the Rust driver's `drivers/rust/src/redwire/scram.rs` byte-for-byte.
//
// Wire shape (RFC 5802):
//   client-first  = "n,," + "n=" + user + ",r=" + cnonce
//   server-first  = "r=" + combined_nonce + ",s=" + b64(salt) + ",i=" + iter
//   client-final  = "c=biws,r=" + combined_nonce + ",p=" + b64(client_proof)
//   server-final  = "v=" + b64(server_signature)
//
// Where:
//   salted_password = PBKDF2-HMAC-SHA256(password, salt, iter, 32)
//   client_key      = HMAC-SHA256(salted_password, "Client Key")
//   stored_key      = SHA256(client_key)
//   client_signature= HMAC-SHA256(stored_key, AuthMessage)
//   client_proof    = client_key XOR client_signature
//   server_key      = HMAC-SHA256(salted_password, "Server Key")
//   server_signature= HMAC-SHA256(server_key, AuthMessage)
//
// AuthMessage = client-first-bare + "," + server-first + "," + client-final-without-proof

const std = @import("std");
const Allocator = std.mem.Allocator;

const HmacSha256 = std.crypto.auth.hmac.sha2.HmacSha256;
const Sha256 = std.crypto.hash.sha2.Sha256;
const pbkdf2 = std.crypto.pwhash.pbkdf2;

pub const HASH_LEN: usize = 32;

pub fn hmacSha256(key: []const u8, data: []const u8) [HASH_LEN]u8 {
    var out: [HASH_LEN]u8 = undefined;
    HmacSha256.create(&out, data, key);
    return out;
}

pub fn sha256(data: []const u8) [HASH_LEN]u8 {
    var out: [HASH_LEN]u8 = undefined;
    Sha256.hash(data, &out, .{});
    return out;
}

/// PBKDF2-HMAC-SHA256 with a fixed 32-byte derived-key length.
pub fn pbkdf2Sha256(password: []const u8, salt: []const u8, iter: u32) ![HASH_LEN]u8 {
    var out: [HASH_LEN]u8 = undefined;
    try pbkdf2(&out, password, salt, iter, HmacSha256);
    return out;
}

pub fn xor(a: []const u8, b: []const u8, dst: []u8) void {
    std.debug.assert(a.len == b.len and dst.len == a.len);
    for (a, b, dst) |x, y, *o| o.* = x ^ y;
}

/// Compute the client proof. Used by the driver to send the
/// `client-final` message — same shape as the engine's
/// `auth/scram.rs::client_proof`.
pub fn clientProof(
    password: []const u8,
    salt: []const u8,
    iter: u32,
    auth_message: []const u8,
) ![HASH_LEN]u8 {
    const salted = try pbkdf2Sha256(password, salt, iter);
    const client_key = hmacSha256(&salted, "Client Key");
    const stored_key = sha256(&client_key);
    const signature = hmacSha256(&stored_key, auth_message);
    var out: [HASH_LEN]u8 = undefined;
    xor(&client_key, &signature, &out);
    return out;
}

pub fn verifyServerSignature(
    password: []const u8,
    salt: []const u8,
    iter: u32,
    auth_message: []const u8,
    presented: []const u8,
) !bool {
    if (presented.len != HASH_LEN) return false;
    const salted = try pbkdf2Sha256(password, salt, iter);
    const server_key = hmacSha256(&salted, "Server Key");
    const expected = hmacSha256(&server_key, auth_message);
    return std.crypto.utils.timingSafeEql([HASH_LEN]u8, expected, presented[0..HASH_LEN].*);
}

// ---------------------------------------------------------------------------
// Helpers used by the handshake state machine. Tiny wrappers around
// std.base64.standard so the call sites stay readable.
// ---------------------------------------------------------------------------

pub const Encoder = std.base64.standard.Encoder;
pub const Decoder = std.base64.standard.Decoder;

pub fn b64Encode(allocator: Allocator, src: []const u8) ![]u8 {
    const out = try allocator.alloc(u8, Encoder.calcSize(src.len));
    errdefer allocator.free(out);
    _ = Encoder.encode(out, src);
    return out;
}

pub fn b64DecodeAlloc(allocator: Allocator, src: []const u8) ![]u8 {
    const max = try Decoder.calcSizeForSlice(src);
    const out = try allocator.alloc(u8, max);
    errdefer allocator.free(out);
    try Decoder.decode(out, src);
    return out;
}

/// Generate a fresh client nonce. RFC 5802 doesn't pin the format —
/// we follow PostgreSQL's convention: 24 random bytes, base64'd
/// without padding-trimming, matching what `examples/stress_wire_client`
/// already produces.
pub fn generateClientNonce(allocator: Allocator) ![]u8 {
    var bytes: [24]u8 = undefined;
    std.crypto.random.bytes(&bytes);
    return b64Encode(allocator, &bytes);
}

/// Parse `r=...,s=...,i=...` out of a server-first message. Returns
/// borrowed slices of the input; combined nonce is returned for the
/// client to echo back in `client-final`.
pub const ServerFirst = struct {
    combined_nonce: []const u8,
    salt_b64: []const u8,
    iter: u32,
};

pub fn parseServerFirst(server_first: []const u8) !ServerFirst {
    var combined_nonce: ?[]const u8 = null;
    var salt_b64: ?[]const u8 = null;
    var iter: ?u32 = null;
    var it = std.mem.splitScalar(u8, server_first, ',');
    while (it.next()) |part| {
        if (part.len < 2) continue;
        if (std.mem.startsWith(u8, part, "r=")) {
            combined_nonce = part[2..];
        } else if (std.mem.startsWith(u8, part, "s=")) {
            salt_b64 = part[2..];
        } else if (std.mem.startsWith(u8, part, "i=")) {
            iter = std.fmt.parseInt(u32, part[2..], 10) catch null;
        }
    }
    if (combined_nonce == null or salt_b64 == null or iter == null) {
        return error.ScramBadServerFirst;
    }
    return ServerFirst{
        .combined_nonce = combined_nonce.?,
        .salt_b64 = salt_b64.?,
        .iter = iter.?,
    };
}

File: drivers/php/src/Redwire/Scram.php
<?php
/**
 * SCRAM-SHA-256 client primitives (RFC 5802 + RFC 7677).
 *
 * Pure functions, no I/O — the state machine that calls these
 * lives in {@see RedwireConn}. Mirrors the engine's
 * `src/auth/scram.rs` and the Java / Rust drivers byte-for-byte.
 */

declare(strict_types=1);

namespace Reddb\Redwire;

use Reddb\RedDBException\ProtocolError;

final class Scram
{
    /** Default iteration count when the server doesn't override. */
    public const DEFAULT_ITER = 16384;
    /** Hard floor — verifiers below this are unsafe to use. */
    public const MIN_ITER = 4096;

    /**
     * Generate 24 random bytes and base64-encode them. Matches the
     * engine-side `new_server_nonce` shape (standard alphabet).
     */
    public static function newClientNonce(): string
    {
        return base64_encode(random_bytes(24));
    }

    /** Build the SCRAM `client-first-message` (no channel binding, no authzid). */
    public static function clientFirst(string $username, string $clientNonce): string
    {
        // GS2 header `n,,` = no channel binding, no authzid.
        return 'n,,n=' . self::saslPrep($username) . ',r=' . $clientNonce;
    }

    /** Strip the `n,,` GS2 header — server-first verification uses the bare form. */
    public static function clientFirstBare(string $clientFirst): string
    {
        if (!str_starts_with($clientFirst, 'n,,')) {
            throw new ProtocolError("client-first must start with 'n,,' (no channel binding)");
        }
        return substr($clientFirst, 3);
    }

    /**
     * Parse `r=<combined>,s=<b64salt>,i=<iter>`. Verifies the
     * combined nonce starts with the client's nonce — replay defence.
     *
     * @return array{combinedNonce: string, salt: string, iter: int, raw: string}
     */
    public static function parseServerFirst(string $serverFirst, string $clientNonce): array
    {
        $combined = null;
        $saltB64 = null;
        $iter = null;
        foreach (explode(',', $serverFirst) as $part) {
            if (str_starts_with($part, 'r=')) {
                $combined = substr($part, 2);
            } elseif (str_starts_with($part, 's=')) {
                $saltB64 = substr($part, 2);
            } elseif (str_starts_with($part, 'i=')) {
                $raw = substr($part, 2);
                if ($raw === '' || !ctype_digit($raw)) {
                    throw new ProtocolError("server-first iter is not an int: {$part}");
                }
                $iter = (int) $raw;
            }
        }
        if ($combined === null) {
            throw new ProtocolError('server-first missing r=');
        }
        if ($saltB64 === null) {
            throw new ProtocolError('server-first missing s=');
        }
        if ($iter === null) {
            throw new ProtocolError('server-first missing i=');
        }
        if (!str_starts_with($combined, $clientNonce)) {
            throw new ProtocolError(
                'server-first nonce does not start with our client nonce — replay protection'
            );
        }
        if ($iter < self::MIN_ITER) {
            throw new ProtocolError(
                "server-first iter {$iter} < MIN_ITER " . self::MIN_ITER
            );
        }
        $salt = base64_decode($saltB64, true);
        if ($salt === false) {
            throw new ProtocolError('server-first salt is not base64');
        }
        return [
            'combinedNonce' => $combined,
            'salt' => $salt,
            'iter' => $iter,
            'raw' => $serverFirst,
        ];
    }

    /** Build the SCRAM `client-final-message-without-proof` (constant `c=biws`). */
    public static function clientFinalNoProof(string $combinedNonce): string
    {
        // c=biws is base64("n,,") — the canonical no-channel-binding header.
        return 'c=biws,r=' . $combinedNonce;
    }

    /** Build the canonical `AuthMessage` per RFC 5802 § 3. */
    public static function authMessage(string $clientFirstBare, string $serverFirst, string $clientFinalNoProof): string
    {
        return $clientFirstBare . ',' . $serverFirst . ',' . $clientFinalNoProof;
    }

    /** PBKDF2-HMAC-SHA256 → 32 bytes. */
    public static function saltedPassword(string $password, string $salt, int $iter): string
    {
        return hash_pbkdf2('sha256', $password, $salt, $iter, 32, true);
    }

    /** HMAC-SHA-256(key, data). */
    public static function hmacSha256(string $key, string $data): string
    {
        return hash_hmac('sha256', $data, $key, true);
    }

    /** SHA-256(data). */
    public static function sha256(string $data): string
    {
        return hash('sha256', $data, true);
    }

    /** Bytewise XOR of two equal-length strings. */
    public static function xor(string $a, string $b): string
    {
        $la = strlen($a);
        $lb = strlen($b);
        if ($la !== $lb) {
            throw new \InvalidArgumentException("xor length mismatch: {$la} vs {$lb}");
        }
        $out = '';
        for ($i = 0; $i < $la; $i++) {
            $out .= chr(ord($a[$i]) ^ ord($b[$i]));
        }
        return $out;
    }

    /**
     * Compute the SCRAM client proof. Mirrors the engine formula:
     * {@code ClientKey XOR HMAC(StoredKey, AuthMessage)}.
     */
    public static function clientProof(string $password, string $salt, int $iter, string $authMessage): string
    {
        $salted = self::saltedPassword($password, $salt, $iter);
        $clientKey = self::hmacSha256($salted, 'Client Key');
        $storedKey = self::sha256($clientKey);
        $sig = self::hmacSha256($storedKey, $authMessage);
        return self::xor($clientKey, $sig);
    }

    /**
     * Verify the server signature. Returns true when the server
     * also knew the verifier (defence against an active MITM).
     * Constant-time comparison via {@see hash_equals()}.
     */
    public static function verifyServerSignature(
        string $password,
        string $salt,
        int $iter,
        string $authMessage,
        string $presentedSignature,
    ): bool {
        if (strlen($presentedSignature) !== 32) {
            return false;
        }
        $salted = self::saltedPassword($password, $salt, $iter);
        $serverKey = self::hmacSha256($salted, 'Server Key');
        $expected = self::hmacSha256($serverKey, $authMessage);
        return hash_equals($expected, $presentedSignature);
    }

    /** Build the SCRAM `client-final-message` (c=,r=,p=). */
    public static function clientFinal(string $combinedNonce, string $proof): string
    {
        return 'c=biws,r=' . $combinedNonce . ',p=' . base64_encode($proof);
    }

    /**
     * Stripped-down SASLprep — RedDB treats usernames as opaque
     * byte strings. We reject `,` and `=` because they break the
     * SCRAM wire format.
     */
    public static function saslPrep(string $input): string
    {
        if (str_contains($input, ',')) {
            throw new \InvalidArgumentException("SCRAM username contains illegal ',': {$input}");
        }
        if (str_contains($input, '=')) {
            throw new \InvalidArgumentException("SCRAM username contains illegal '=': {$input}");
        }
        return $input;
    }
}
